src/hooks/useModifyCharacters.ts

```diff
@@ -2,44 +2,196 @@ import { useReducer } from 'react'
 
 import { ModifiedCharacter } from '../api/starWars-types'
 
+export const CHARACTER_ACTION_TYPES = {
+  SORT: 'sort',
+  FILTER: 'filter',
+} as const
+
 export const SORT_FACTOR = {
   MASS: 'mass',
   HEIGHT: 'height',
 } as const
 
+export const FILTER_FACTOR = {
+  HOMEWORLD: 'homeworld',
+  GENDER: 'gender',
+} as const
+
 // "mass" | "height"
 export type SortFactor = typeof SORT_FACTOR[keyof typeof SORT_FACTOR]
 
+export type FilterFactor = typeof FILTER_FACTOR[keyof typeof FILTER_FACTOR]
+
+type FilterTargets = {
+  homelandTarget: string
+  genderTarget: string
+}
+
 type CharacterState = {
   originalCharacters: ModifiedCharacter[]
-  modifiedCharacters: ModifiedCharacter[]
-  selectedValue: SortFactor | ''
+  renderedCharacters: ModifiedCharacter[]
+  sortFactor: SortFactor | ''
+  homeworldFilters: string[]
+  genderFilters: string[]
+  filterTargets: FilterTargets
 }
 
-type CharacterAction = { type: 'sort'; sortFactor: CharacterState['selectedValue'] }
+type CharacterAction =
+  | {
+      type: 'sort'
+      sortFactor: CharacterState['sortFactor']
+    }
+  | { type: 'filter'; filterFactor: FilterFactor; filterTarget: string }
 
 const sort = (characters: ModifiedCharacter[], key: SortFactor) =>
   characters.sort((a, b) => (b[key] || 0) - (a[key] || 0))
 
+const filter = (characters: ModifiedCharacter[], key: FilterFactor, filterTarget: string) =>
+  characters.filter((character) => character[key] === filterTarget)
+
+const hasTarget = (
+  filterFactor: FilterFactor,
+  previousHomelandTarget: string,
+  previousGenderTarget: string,
+) => {
+  let hasTarget = false
+  if (filterFactor === FILTER_FACTOR.HOMEWORLD && previousHomelandTarget !== '') {
+    hasTarget = true
+  }
+  if (filterFactor === FILTER_FACTOR.GENDER && previousGenderTarget !== '') {
+    hasTarget = true
+  }
+  return hasTarget
+}
+
 const characterReducer = (state: CharacterState, action: CharacterAction): CharacterState => {
   switch (action.type) {
-    case 'sort':
+    case CHARACTER_ACTION_TYPES.SORT: {
       if (action.sortFactor === 'height' || action.sortFactor === 'mass') {
-        const newModifiedCharacters = [...state.modifiedCharacters]
-        sort(newModifiedCharacters, action.sortFactor)
+        const newRenderedCharacters = [...state.renderedCharacters]
+        sort(newRenderedCharacters, action.sortFactor)
+        return {
+          ...state,
+          renderedCharacters: newRenderedCharacters,
+          sortFactor: action.sortFactor,
+        }
+      }
+      const homelandTarget = state.filterTargets.homelandTarget
+      const genderTarget = state.filterTargets.genderTarget
+      let newRenderedCharacters = state.originalCharacters
```

Comment:
These lines are too close to each other that it's harming readability a bit. prob you can do like;

```suggestion
        const newRenderedCharacters = [...state.renderedCharacters]
        sort(newRenderedCharacters, action.sortFactor)
        
        return {
          ...state,
          renderedCharacters: newRenderedCharacters,
          sortFactor: action.sortFactor,
        }
      }
      
      const homelandTarget = state.filterTargets.homelandTarget
      const genderTarget = state.filterTargets.genderTarget
      let newRenderedCharacters = state.originalCharacters
```